apis/apiextensions/v1/composition_types.go

```diff
@@ -349,6 +425,42 @@ func (c *Patch) applyFromFieldPathPatch(from, to runtime.Object) error { // noli
 	return runtime.DefaultUnstructuredConverter.FromUnstructured(toMap, to)
 }
 
+// applyFromFnPatch patches the "to" resource using a function defined in
+// the FromFn. Values may be transformed if any are defined on
+// the patch.
+func (c *Patch) applyFromFnPatch(to runtime.Object) error {
+	if c.ToFieldPath == nil {
+		return errors.Errorf(errRequiredField, "ToFieldPath", c.Type)
+	}
+
+	if c.FromFn == nil {
+		return errors.Errorf(errRequiredField, "FromFn", c.Type)
+	}
+
+	fnValue, err := c.FromFn.GetFnValue()
+	if err != nil {
+		return err
+	}
+
+	out := fnValue
+	for i, f := range c.Transforms {
+		if out, err = f.Transform(out); err != nil {
+			return errors.Wrapf(err, errFmtTransformAtIndex, i)
+		}
+	}
+
+	toMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(to)
```

Comment:
I think with this PR and https://github.com/crossplane/crossplane/pull/2298, we are repeating this code pattern. I wonder if we have an opportunity to refactor out the field mapping & conversion (and maybe document what is happening internally a bit bette).